pkg/pillar/cmd/zedrouter/ipaddronunet.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Network Instance underlay network Application Number Management
+
+package zedrouter
+
+import (
+	"fmt"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+func appNumsOnUNetAllocate(ctx *zedrouterContext,
+	config *types.AppNetworkConfig) {
+	for _, ulConfig := range config.UnderlayNetworkList {
+		appID := config.UUIDandVersion.UUID
+		networkID := ulConfig.Network
+		netStatus := lookupNetworkInstanceStatus(ctx, networkID.String())
+		if netStatus != nil {
+			baseSize := netStatus.DhcpRange.Size
+			appNum, err := appNumOnUNetAllocate(ctx, networkID, appID,
+				baseSize, false)
+			if appNum == types.AppNumInvalid || err != nil {
+				errStr := fmt.Sprintf("App Num get fail :%s", err)
+				log.Errorf("appNumsOnAppNetworkAllocate(%s, %s): fail: %s\n",
+					networkID.String(), appID.String(), errStr)
+			}
+		}
+	}
+}
+
+func appNumsOnUNetFree(ctx *zedrouterContext,
+	status *types.AppNetworkStatus) {
+	appID := status.UUIDandVersion.UUID
+	for ulNum := 0; ulNum < len(status.UnderlayNetworkList); ulNum++ {
+		ulStatus := &status.UnderlayNetworkList[ulNum]
+		networkID := ulStatus.Network
+		// release the app number
```

Comment:
How do you know it was allocated?
I think it makes sense to have appNumsOnUNetAllocate() to save the appnumonunet in the underlayNetworkStatus so you can check that here before potentially freeing something which wasn't allocated, and also access that field instead of a Get function.